pkg/pillar/nireconciler/genericitems/dnsmasq.go

```diff
@@ -606,24 +626,57 @@ func (c *DnsmasqConfigurator) CreateDnsmasqConfig(buffer io.Writer, dnsmasq Dnsm
 		}
 	}
 
-	if router != "" {
+	// Prepare the set of all static routes to propagate to applications.
+	var staticRoutes []types.IPRoute
+	if !isIPv6 {
+		if dnsmasq.DHCPServer.AllOnesNetmask {
+			// Make the gateway reachable using a link-local route.
+			if gatewayIP != nil {
+				staticRoutes = append(staticRoutes, types.IPRoute{
+					DstNetwork: netutils.HostSubnet(gatewayIP),
+					Gateway:    net.IP{0, 0, 0, 0},
+				})
+			}
+			if subnet != nil && gatewayIP != nil {
+				staticRoutes = append(staticRoutes, types.IPRoute{
+					DstNetwork: subnet,
+					Gateway:    gatewayIP,
+				})
+			}
+		} else {
+			// With all-ones netmask disabled we use forwarding between apps
+			// on the same NI.
+			if subnet != nil {
+				staticRoutes = append(staticRoutes, types.IPRoute{
+					DstNetwork: subnet,
+					Gateway:    net.IP{0, 0, 0, 0},
+				})
+			}
+		}
+		staticRoutes = append(staticRoutes, dnsmasq.DHCPServer.PropagateRoutes...)
+	}
+
+	if gatewayIP != nil && dnsmasq.DHCPServer.WithDefaultRoute {
 		// XXX IPv6 needs to be handled in radvd.
 		if !isIPv6 {
 			if _, err := io.WriteString(buffer,
 				fmt.Sprintf("dhcp-option=option:router,%s\n",
 					dnsmasq.DHCPServer.GatewayIP)); err != nil {
 				return writeErr(err)
 			}
-			if dnsmasq.DHCPServer.AllOnesNetmask {
-				if _, err := io.WriteString(buffer,
-					fmt.Sprintf(
-						"dhcp-option=option:classless-static-route,%s/32,%s,%s,%s,%s,%s\n",
-						router, "0.0.0.0",
-						"0.0.0.0/0", router,
-						subnet, router)); err != nil {
-					return writeErr(err)
-				}
-			}
+			// From RFC 3442:
+			// If the DHCP server returns both a Classless Static Routes option and
+			// a Router option, the DHCP client MUST ignore the Router option.
+			//
+			// This means that we have to include a default "classless" route
+			// even though the Router option is provided.
+			// We keep the Router option for clients that do not support the Classless
```

Comment:
Makes sense to test this with a range of guest VMs including some version(s) of Windows, to make sure no client is upset by seeing both options.

Reply:
Will do

Reply:
Tested with Windows-es - looks good, IP routes are installed as expected.